x-pack/plugins/ingest_manager/server/services/epm/packages/install.ts

```diff
@@ -95,6 +99,117 @@ export async function ensureInstalledPackage(options: {
   return installation;
 }
 
+export async function handleInstallPackageFailure(
+  savedObjectsClient: SavedObjectsClientContract,
+  error: IngestManagerError | Boom | Error,
+  pkgName: string,
+  pkgVersion: string,
+  installedPkg: SavedObject<Installation> | undefined,
+  callCluster: CallESAsCurrentUser
+) {
+  if (error instanceof IngestManagerError) {
+    return;
+  }
+  const logger = appContextService.getLogger();
+  const pkgkey = Registry.pkgToPkgKey({
+    name: pkgName,
+    version: pkgVersion,
+  });
+
+  // if there is an unknown server error, uninstall any package assets or reinstall the previous version if update
+  try {
+    const installType = getInstallType({ pkgVersion, installedPkg });
+    if (installType === 'install' || installType === 'reinstall') {
+      logger.error(`uninstalling ${pkgkey} after error installing`);
+      await removeInstallation({ savedObjectsClient, pkgkey, callCluster });
+    }
+
+    if (installType === 'update') {
+      if (!installedPkg) {
+        logger.error(
+          `failed to rollback package after installation error ${error} because saved object was undefined`
+        );
+        return;
+      }
+      const prevVersion = `${pkgName}-${installedPkg.attributes.version}`;
+      logger.error(`rolling back to ${prevVersion} after error installing ${pkgkey}`);
+      await installPackage({
+        savedObjectsClient,
+        pkgkey: prevVersion,
+        callCluster,
+      });
+    }
+  } catch (e) {
+    logger.error(`failed to uninstall or rollback package after installation error ${e}`);
+  }
+}
+
+export async function upgradePackages({
+  savedObjectsClient,
+  packagesToUpgrade,
+  callCluster,
+}: {
+  savedObjectsClient: SavedObjectsClientContract;
+  packagesToUpgrade: string[];
+  callCluster: CallESAsCurrentUser;
+}): Promise<Array<UpgradePackageInfo | UpgradePackageError>> {
+  const res: Array<UpgradePackageInfo | UpgradePackageError> = [];
+
+  for (const pkgToUpgrade of packagesToUpgrade) {
```

Comment:
Since we don't care want to stop on the first failure, I think we can use Promise.allSettled https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled and loop the results and use the `status` property to test success/failure

Reply:
Ah thanks!

Reply:
John and I talked about this and I think we're going to leave it as is for now because the code I came up with using `allSettled` wasn't super readable.

Reply:
Thanks John! I'll pull in your changes.